src/dataframe.rs

```diff
@@ -451,6 +454,26 @@ impl PyDataFrame {
         Ok(table)
     }
 
+    #[allow(unused_variables)]
+    fn __arrow_c_stream__<'py>(
+        &'py mut self,
+        py: Python<'py>,
+        requested_schema: Option<Bound<'py, PyCapsule>>,
+    ) -> PyResult<Bound<'py, PyCapsule>> {
+        let batches = wait_for_future(py, self.df.as_ref().clone().collect())?
+            .into_iter()
+            .map(|r| Ok(r));
```

Comment:
It would be cool if we didn't have to collect all the batches ahead of time, but could synchronously collect one batch at a time within the iterator. I have no idea if datafusion would support that though.

Reply:
Yeah, I was wondering the same. I think it would mean creating a `TableProvider` for these, but that's a heavier lift than I wanted to get into right now.